code/framework/api/src/main/java/io/cattle/platform/api/resource/jooq/AbstractJooqResourceManager.java

```diff
@@ -85,11 +91,46 @@ protected Object listInternal(SchemaFactory schemaFactory, String type, Map<Obje
 
         List<?> result = mapper == null ? query.fetch() : query.fetchInto(mapper);
 
+        addCapabilities(result, schemaFactory.getSchema(clz));
+
         processPaginationResult(result, pagination, mapper);
 
         return result;
     }
 
+    private void addCapabilities(List<?> result, Schema s) {
+        // This could be built into a framework of "capability providers" where the providers are stored in a map with the key being the 
+        // capability, but since we only have one for now, we'll do it the simple way
+        if (s == null) {
+            return;
+        }
+
+        Map<String, Action> actions = s.getResourceActions();
+        if (actions == null || actions.size() == 0) {
+            return;
+        }
+
+        boolean addCapability = false;
+        for (Map.Entry<String, Action> entry : actions.entrySet()) {
+            Map<String, Object> attributes = entry.getValue().getAttributes();
+            if (attributes == null || attributes.size() == 0) {
+                continue;
+            }
+
+            String capability = ObjectUtils.toString(attributes.get(ObjectMetaDataManager.CAPABILITY), null);
+            if (ObjectMetaDataManager.MODIFY_INFRA_CAPABILITY.equals(capability)) {
+                addCapability = true;
+                break;
+            }
+        }
+
+        if (addCapability && infraAccess.canModifyInfrastructure((Policy) ApiContext.getContext().getPolicy())) {
```

Comment:
Nit: can the check for infraAccess.canModifyInfrastructure((Policy) ApiContext.getContext().getPolicy()) be moved up, so when it is false, we can return immediately? This way we can save a bit on going over capabilities.